balancer/pickfirst/pickfirstleaf/pickfirstleaf.go

```diff
@@ -384,8 +405,10 @@ func (b *pickfirstBalancer) requestConnectionLocked() {
 		switch scd.state {
 		case connectivity.Idle:
 			scd.subConn.Connect()
+			b.scheduleNextConnectionLocked()
 		case connectivity.TransientFailure:
 			// Try the next address.
+			scd.connectionAttempted = true
```

Comment:
The connection attempted flag is reset when the first pass is started. A re-used subchannel may not have completed backoff when the attempt is made. This case is mentioned in [A62](https://github.com/grpc/proposal/blob/master/A61-IPv4-IPv6-dualstack-backends.md#happy-eyeballs-in-the-pick_first-lb-policy):

> Note that every time the LB policy receives a new address list, it will start an initial Happy Eyeballs pass over the new list, even if some of the subchannels are not actually new due to their addresses having been present on both the old and new lists. This means that on the initial pass through the address list for a subsequent address list update, when pick_first decides to start a connection attempt on a given subchannel (whether because it is the first subchannel in the list or because the timer fired before the previous address' connection attempt completed), that subchannel may not be in state IDLE, which is the only state in which a connection attempt may be requested.

and later:
> * If the subchannel is in state TRANSIENT_FAILURE, then we know that it is in backoff due to a recent connection attempt failure, so we treat it as if we have already made a connection attempt on this subchannel, and we will immediately move on to the next subchannel.

Reply:
Thanks. Makes sense.